source/hwIo/ble/_io.py

```diff
@@ -0,0 +1,233 @@
+
+
+def sliced(data: bytes, n: int) -> Iterator[bytes]:
+	"""Split data into chunks of size n (last chunk may be smaller)."""
+	return takewhile(len, (data[i : i + n] for i in count(0, n)))
+
+
+class Ble(IoBase):
+	"""I/O for Bluetooth Low Energy (BLE) devices
+
+	This implementation expects a service/characteristic pair to send raw data to as a BLE command
+	and receive raw data through a BLE notify on a service/characteristic pair.
+	"""
+
+	_client: bleak.BleakClient
+	"The Bleak client to use for BLE communication"
+	_writeServiceUuid: str
+	"The service UUID to use for writing data to the peripheral, this should accept BLE commands"
+	_writeCharacteristicUuid: str
+	"The characteristic UUID to use for writing data to the peripheral, this should accept BLE commands"
+	_readServiceUuid: str
+	"The service UUID to use for reading data from the peripheral, this should generate BLE notifications"
+	_readCharacteristicUuid: str
+	"""The characteristic UUID to use for reading data from the peripheral,
+	this should generate BLE notifications"""
+	_onReceive: Callable[[bytes], None] | None
+	"The callback to call when data is received"
+	_queuedData: Queue[bytes | bytearray]
+	"A queue of received data, this is processsed by the onReceive handler"
+	_readEvent: Event
+	"An event that is set when data is received"
+	_readerThread: Thread
+	"Thread that processes the queue of rad data"
+	_stopReaderEvent: Event
+	"Event that is set to stop the reader thread"
+	_ioThreadRef: weakref.ReferenceType[IoThread]
+	"Reference to the I/O thread"
+
+	def __init__(
+		self,
+		device: BLEDevice | str,
+		writeServiceUuid: str,
+		writeCharacteristicUuid: str,
+		readServiceUuid: str,
+		readCharacteristicUuid: str,
+		onReceive: Callable[[bytes], None],
+		ioThread: IoThread | None = None,
+	) -> None:
+		if isinstance(device, str):
+			# String address provided - Bleak will perform implicit discovery
+			address = device
+			log.info(f"Connecting to BLE device at address {address}")
+			self._client = bleak.BleakClient(address, winrt=WINRT_CLIENT_ARGS)
+		else:
+			# BLEDevice object provided (preferred)
+			log.info(f"Connecting to {device.name} ({device.address})")
+			self._client = bleak.BleakClient(device, winrt=WINRT_CLIENT_ARGS)
+		self._writeServiceUuid = writeServiceUuid
+		self._writeCharacteristicUuid = writeCharacteristicUuid
+		self._readServiceUuid = readServiceUuid
+		self._readCharacteristicUuid = readCharacteristicUuid
+		self._onReceive = onReceive
+		if ioThread is None:
+			from .. import bgThread as ioThread
+		self._ioThreadRef = weakref.ref(ioThread)
+		self._queuedData = Queue()
+		self._readEvent = Event()
+		self._stopReaderEvent = Event()
+		self._readerThread = Thread(
+			target=queueReader,
+			args=(self._queuedData, self._onReceive, self._stopReaderEvent, ioThread),
+			daemon=True,
+		)
+		self._readerThread.start()
+		f = runCoroutine(self._initAndConnect())
+		f.result()
```

Comment:
Could you consider creating a wrapper for this in `asyncioEventLoop`? E.g. a function that receives the coroutine, runs it, fetches the result, raises the exception when it fails.